### PR TITLE
Feature/common indexer

### PR DESCRIPTION
**Pull request by payonear**
Common indexer added for BPE, BERT and Sentence Transformer into text_utils folder.
Test procedures for old and new indexers added to test folder.
Switched to transformers from pytorch_pretrained_bert.
Old model scripts moved to ml_prev folder.
New training dataset created (details in `know_base.ipynb`)

### CODE REVIEW

File: evaluate_models.py [hunk 1]
```diff
@@ -35,17 +49,20 @@ def print_scores(df):
         model_ans_list = []
         true_ans_list = []
```
Comment (submaps) [inline evaluate_models.py:50]: true_ans_list is not necessary longer

File: evaluate_models.py [hunk 2]
```diff
@@ -35,17 +49,20 @@ def print_scores(df):
...
-            if links:
-                links = [x.strip().split('#')[0] for x in links]
-                model_ans_list.append(links[0])
-                acc_top_1 = int(true_link == links[0])
-                acc_top_3 = int(any(true_link == link for link in links[:3]))
+            true_ans_list.append(true_ans)
+            # print('\t\t', true_ans[:100], ans_list[0][:100])
+            if len(ans_list) > 0:
+                model_ans_list.append(ans_list)
+                acc_top_1 = int(is_equal(true_ans, ans_list[0]))
```
Comment (submaps) [inline evaluate_models.py:60]: All metrics (acc_top_1, acc_top_3 metrics and future) it's better to extract to one object and method to calculate
`metric = get_metric(true_ans, ans_list)`
and further save it

File: evaluate_models.py [hunk 3]
```diff
@@ -54,30 +71,34 @@ def print_scores(df):
...
+        # mtx_df.to_csv(f'output/conf_mtx_{model_name}.csv')
+        # print(mtx_df.shape)
+        # print(mtx_df)
         # print(model_name.value, score_top_1 / n, score_top_3 / n, np.mean(elapsed_list))
 
     report = pd.DataFrame(lines)
     return report
 
 
 for slice_name, df_slice in df.groupby('source'):
```
Comment (submaps) [inline evaluate_models.py:93]: To avoid silent execution during static import it's better to extract these code to main function 
`if __name__ == '__main__':
     main()`

File: evaluate_models.py [hunk 4]
```diff
@@ -54,30 +71,34 @@ def print_scores(df):
...
-    df_slice = df_slice[['questions', 'url_3', 'section_1']].dropna()
-    if len(df_slice) > 5:
-        report = print_scores(df_slice)
-        report.to_csv(f'output/report_{slice_name}.csv', index=False)
-        print(slice_name)
-        print(report)
+    print('start score for slice ', slice_name)
+    print('data shape: ', df_slice.shape)
+    report = print_scores(questions, true_answers)
+    report.to_csv(f'{DATA}/output/report_{slice_name}.csv', index=False)
```
Comment (submaps) [inline evaluate_models.py:102]: All file paths it should be placed at the top of file, after saving we can print a message `saved to $file_path`

File: evaluate_models.py [hunk 5]
```diff
@@ -1,30 +1,44 @@
...
-    new_order = sorted(new_labels)
-    df_mtx = df_mtx.loc[new_order, new_order]
-    return df_mtx
+def is_equal(true_ans, pred_ans):
+    return pred_ans in true_ans or true_ans in pred_ans
+           # or nltk.edit_distance(true_ans, pred_ans) / len(true_ans) < 0.4
 
 
-def print_scores(df):
+def print_scores(questions, true_answers):
```
Comment (submaps) [inline evaluate_models.py:41]: The function name is confusing, maybe it should be renamed `evaluate_models`

File: evaluate_models.py [hunk 6]
```diff
@@ -35,17 +49,20 @@ def print_scores(df):
         model_ans_list = []
         true_ans_list = []
 
-        for q, true_link in zip(df['questions'].values, df['url_3'].values):
+        for q, true_ans in zip(questions, true_answers):
```
Comment (submaps) [inline evaluate_models.py:52]: Too many nested circles, I suggest to replace inner circle to `evaluate_model` method

File: ml_models/bert_model.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
+from transformers import BertModel, BertTokenizer
+from functools import wraps
+import numpy as np
+from text_utils.indexer import *
```
Comment (submaps) [inline ml_models/bert_model.py]: bad practice

File: ml_models/bert_model.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    return inner
+
+
+class BertEmbedder:
+    """
+    Embedding Wrapper on Bert Multilingual Uncased
+    """
+
+    def __init__(self):
+        self.model_file = 'bert-base-multilingual-uncased'
```
Comment (submaps) [inline ml_models/bert_model.py:33]: Possible we can add `model_file`, `vocab_file` settings as default variables to constructor

File: ml_models/bert_model.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        self.error_count = 0
+
+    @singleton
+    def bert_model(self):
+        model = BertModel.from_pretrained(self.model_file).eval()
+        return model
+
+    @singleton
+    def bert_tokenizer(self):
+        tokenizer = BertTokenizer.from_pretrained(self.vocab_file, do_lower_case=True)
```
Comment (submaps) [inline ml_models/bert_model.py:47]: `do_lower_case=True` parameter depends to using model `bert-base-multilingual-uncased`, it's better to replace this param to constructor and add a checking condition

File: ml_models/bert_model.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+                        truncation=True,
+                        return_tensors = 'pt').values()
+            
+            with torch.no_grad():
+                encoded_layers, _ = self.model(inputs_ids, attention_mask=attention_mask, token_type_ids=token_type_ids)
+            sent_embedding = encoded_layers.mean(dim=1)
+            vect = sent_embedding[0].numpy()
+            self.success_count += 1
+            return vect
+        except:
```
Comment (submaps) [inline ml_models/bert_model.py:64]: we should log exceptions

File: ml_models/bert_model.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+
+def get_answer(query):
+    ans_list = [get_text_by_ind(ind) for k, ind in indexer.return_closest(query, k=4)]
+    return ans_list
+
+logger.info('bert indexer started')
+indexer, df = prepare_indexer('bert', logger)
+logger.info('bert indexer ready')
+check_indexer()
```
Comment (submaps) [inline ml_models/bert_model.py]: add new empty line to the end of file

File: ml_models/bpe_model.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
+from functools import wraps
+import numpy as np
+from text_utils.indexer import *
```
Comment (submaps) [inline ml_models/bpe_model.py]: use direct import

File: ml_models/bpe_model.py [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+    @wraps(cls)
+    def inner(*args, **kwargs):
+        nonlocal instance
+        if instance is None:
+            instance = cls(*args, **kwargs)
+        return instance
+
+    return inner
+
+class BPEEmbedder:
```
Comment (submaps) [inline ml_models/bpe_model.py:24]: Possible we should make a base interface to these models, and then implement it
```
import abc
class BasicEmbedder(abc.ABC):
    def sentence_embedding(text):
        pass
```

File: ml_models/bpe_model.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+
+
+def get_answer(query):
+    ans_list = [get_text_by_ind(ind) for k, ind in indexer.return_closest(query, k=4)]
+    return ans_list
+
+logger.info('bpe indexer started')
+indexer, df = prepare_indexer('bpe', logger)
+logger.info('bpe indexer ready')
+check_indexer()
```
Comment (submaps) [inline ml_models/bpe_model.py]: last \n

File: ml_models/bpe_model.py [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+        try:
+            mean_emb = self.model.embed(text).mean(axis=0)
+            self.success_count += 1
+            return mean_emb / np.linalg.norm(mean_emb)
+        except:
+            self.error_count += 1
+        return np.zeros(FEATURE_SIZE)
+
+def check_indexer():
+    test_queries = ['Есть ли аналоги pandas (ну или не аналоги а тоже либы для работы с данными) для работы с данными',
```
Comment (submaps) [inline ml_models/bpe_model.py:45]: Duplicated code, it can be replaced to new file `examples.py`

File: ml_models/use_model.py [hunk 16]
```diff
@@ -0,0 +1,76 @@
+from sentence_transformers import SentenceTransformer
+from functools import wraps
+import numpy as np
+from text_utils.indexer import *
```
Comment (submaps) [inline ml_models/use_model.py]: need to fix)

File: ml_models/use_model.py [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+        try:
+            sent_embedding = self.model.encode(text)
+            self.success_count += 1
+            return sent_embedding
+        except:
+            self.error_count += 1
+        return np.zeros(FEATURE_SIZE)
+
+def check_indexer():
+    test_queries = ['Есть ли аналоги pandas (ну или не аналоги а тоже либы для работы с данными) для работы с данными',
```
Comment (submaps) [inline ml_models/use_model.py:50]: import from `examples.py`

File: ml_models/use_model.py [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+
+
+def get_answer(query):
+    ans_list = [get_text_by_ind(ind) for k, ind in indexer.return_closest(query, k=4)]
+    return ans_list
+
+logger.info('use indexer started')
+indexer, df = prepare_indexer('use', logger)
+logger.info('use indexer ready')
+check_indexer()
```
Comment (submaps) [inline ml_models/use_model.py]: last \n

File: test/test_indexer_new.py [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+
+
+    def test_use_indexer(self):
+        from ml_models.use_model import indexer
+        for q in questions:
+            r = indexer.model.sentence_embedding(q)
+            self.assertTrue(r.shape[0]==768) # embedding dimension
+
+            res = indexer.index.knnQueryBatch(queries=[r], k=4, num_threads=2)
+            self.assertTrue(len(res[0][0])==4) # return 4 neighbours
```
Comment (submaps) [inline test/test_indexer_new.py]: can be used assertEqual

File: test/test_indexer_new.py [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+            self.assertTrue(res[0][1][0]<1e-5) # find itself, distance apx. 0
```
Comment (submaps) [inline test/test_indexer_new.py]: can be used assertLess

File: text_utils/indexer.py [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+        else:
+            raise ValueError(f'Model {model_name} not found.')
+        self.model_name = model_name
+        self.space_type = 'cosinesimil'
+        self.method_name = 'hnsw'
+        self.index = nmslib.init(space=self.space_type,
+                                 method=self.method_name,
+                                 data_type=nmslib.DataType.DENSE_VECTOR,
+                                 dtype=nmslib.DistType.FLOAT)
+        self.index_params = {'NN': 15}
```
Comment (submaps) [inline text_utils/indexer.py]: It does not use

File: text_utils/utils.py [hunk 22]
```diff
@@ -1,17 +1,19 @@
...
     logger.setLevel(logging.DEBUG)
-    formatter = logging.Formatter('%(asctime)s %(message)s')
+    formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
     file_handler = logging.FileHandler(filename=logger_path)
     file_handler.setFormatter(formatter)
     stream_handler = logging.StreamHandler()
     stream_handler.setFormatter(formatter)
     logger.addHandler(stream_handler)
     logger.addHandler(file_handler)
+    os.makedirs('logs', exist_ok=True)
```
Comment (submaps) [inline text_utils/utils.py:16]: How to be sure we in the right dir? It makes directory related to current importing script path

**Review by submaps — COMMENTED** [review]
Add comments about class structure and codestyle, requiring minor refactoring changes.

File: evaluate_models.py (hunk 5 above)
Comment (payonear) [reply evaluate_models.py:41]: To be fixed in separate patch.

File: evaluate_models.py (hunk 1 above)
Comment (payonear) [reply evaluate_models.py:50]: To be fixed in separate patch.

File: evaluate_models.py (hunk 6 above)
Comment (payonear) [reply evaluate_models.py:52]: To be fixed in separate patch.

File: evaluate_models.py (hunk 2 above)
Comment (payonear) [reply evaluate_models.py:60]: To be fixed in separate patch.

File: evaluate_models.py (hunk 3 above)
Comment (payonear) [reply evaluate_models.py:93]: To be fixed in separate patch.

File: evaluate_models.py (hunk 4 above)
Comment (payonear) [reply evaluate_models.py:102]: To be fixed in separate patch.

File: ml_models/bert_model.py (hunk 7 above)
Comment (payonear) [reply ml_models/bert_model.py]: fixed

File: ml_models/bert_model.py (hunk 8 above)
Comment (payonear) [reply ml_models/bert_model.py:33]: To be fixed in separate patch.

File: ml_models/use_model.py (hunk 16 above)
Comment (payonear) [reply ml_models/use_model.py]: fixed

File: ml_models/use_model.py (hunk 18 above)
Comment (payonear) [reply ml_models/use_model.py]: fixed

File: ml_models/bert_model.py (hunk 11 above)
Comment (payonear) [reply ml_models/bert_model.py]: fixed

File: ml_models/bpe_model.py (hunk 12 above)
Comment (payonear) [reply ml_models/bpe_model.py]: fixed

File: ml_models/bpe_model.py (hunk 13 above)
Comment (payonear) [reply ml_models/bpe_model.py:24]: To be implemented in separate patch.

File: ml_models/bpe_model.py (hunk 15 above)
Comment (payonear) [reply ml_models/bpe_model.py:45]: To be fixed in separate patch.

File: ml_models/bpe_model.py (hunk 14 above)
Comment (payonear) [reply ml_models/bpe_model.py]: fixed

File: ml_models/use_model.py (hunk 17 above)
Comment (payonear) [reply ml_models/use_model.py:50]: To be fixed in separate patch.

File: test/test_indexer_new.py (hunk 19 above)
Comment (payonear) [reply test/test_indexer_new.py]: fixed

File: test/test_indexer_new.py (hunk 20 above)
Comment (payonear) [reply test/test_indexer_new.py]: fixed

File: text_utils/utils.py (hunk 22 above)
Comment (payonear) [reply text_utils/utils.py:16]: To be fixed in separate patch

File: ml_models/bert_model.py (hunk 10 above)
Comment (payonear) [reply ml_models/bert_model.py:64]: To be implemented in separate patch.

File: ml_models/bert_model.py (hunk 9 above)
Comment (payonear) [reply ml_models/bert_model.py:47]: To be implemented in separate patch.

File: text_utils/indexer.py (hunk 21 above)
Comment (payonear) [reply text_utils/indexer.py]: fixed

Comment (submaps) [reply text_utils/indexer.py]: Not found a fix at the new branch
